src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -32,37 +50,90 @@ public abstract class AssemblyLoadContext
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppContext.Unloading
+            AppContext.Unloading += OnAppContextUnloading;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+        protected AssemblyLoadContext() : this(false, false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
 
-        private void InitializeLoadContext(bool fRepresentsTPALoadContext)
+        protected AssemblyLoadContext(bool isCollectible) : this(false, isCollectible)
+        {
+        }
+
+        internal AssemblyLoadContext(bool fRepresentsTPALoadContext, bool isCollectible)
         {
             // Initialize the VM side of AssemblyLoadContext if not already done.
-            GCHandle gchALC = GCHandle.Alloc(this);
-            IntPtr ptrALC = GCHandle.ToIntPtr(gchALC);
-            m_pNativeAssemblyLoadContext = InitializeAssemblyLoadContext(ptrALC, fRepresentsTPALoadContext);
+            IsCollectible = isCollectible;
+
+            // Add this instance to the list of alive ALC
+            lock (ContextsToUnload)
+            {
+                if (_isProcessExiting)
+                {
+                    throw new InvalidOperationException(SR.GetResourceString("AssemblyLoadContext_Constructor_CannotInstantiateWhileUnloading"));
+                }
+
+                // If this is a collectible ALC, we are creating a weak handle otherwise we use a strong handle
+                var thisHandle = GCHandle.Alloc(this, IsCollectible ? GCHandleType.Weak : GCHandleType.Normal);
+                var thisHandlePtr = GCHandle.ToIntPtr(thisHandle);
+                m_pNativeAssemblyLoadContext = InitializeAssemblyLoadContext(thisHandlePtr, fRepresentsTPALoadContext, isCollectible);
+
+                // Initialize event handlers to be null by default
+                Resolving = null;
+                Unloading = null;
+
+                id = _nextId++;
+                ContextsToUnload.Add(id, new WeakReference<AssemblyLoadContext>(this, true));
+            }
+        }
+
+        ~AssemblyLoadContext()
+        {
+            // Only valid for a Collectible ALC. Non-collectible ALCs have the finalizer suppressed.
+            // We get here only in case the explicit Unload was not initiated.
+            Debug.Assert(state != InternalState.Unloading);
+            InitiateUnload();
+        }
 
-            // Initialize event handlers to be null by default
-            Resolving = null;
+        private void InitiateUnload()
+        {
+            var unloading = Unloading;
             Unloading = null;
+            unloading?.Invoke(this);
```

Comment:
This may be launching code inside of the finalizer thread. Do we think that's ok?